### PR TITLE
refactor(1789): VisionLanguage + Document + NER + ComputerVision (09/18)

### PR DESCRIPTION
**Pull request by ooples**
Slice 9 of 18 splitting #1789 into review-sized PRs.

#1789 reached **1,720 changed files**, so it has never received an automated review. CodeRabbit caps reviews at **100 changed files** and rate-limits, so the split is packed close to the cap: 18 slices of 95-96 files each, which is the minimum PR count possible.

**Scope:** VisionLanguage + Document + NER + ComputerVision

**Files:** 96

## Why an integration branch

These slices target `integration/1789`, not `master`. The test generator and the model fixes are coupled in both directions - the generator's smoke constructors are what surfaced the model bugs, and its `Fp32` / `HeavyTimeout` mitigation lists only make sense once those models are fixed. Making each slice independently green against `master` would mean re-verifying the full shard matrix 18 times with red intermediate states. Each slice is reviewed at reviewable size; `integration/1789` is what must be green before it merges to `master`.

## Integrity

The 18 slices were cut by file scope from a single snapshot and verified mechanically: their union is exactly the 1,720 files, with **zero overlaps and zero gaps**, and every slice is **byte-identical** to the snapshot for the files it carries.

## Individual slices do not build standalone

Expected under this model - the slices are cut by scope, not by compilability, and cross-reference each other. `integration/1789` is the unit that must build and go green.

Related: #1789

### CODE REVIEW

File: src/Document/VisionLanguage/UDOP.cs [hunk 1]
```diff
@@ -252,6 +308,35 @@ protected override void InitializeLayers()
...
+    {
+        if (_lazyShapesWarmed) return;
+        _lazyShapesWarmed = true;
+        if (!_useNativeMode) return;
+
+        var dummy = new Tensor<T>([3, ImageSize, ImageSize]);
+        bool wasTraining = IsTrainingMode;
+        if (wasTraining) SetTrainingMode(false);
+        try { _ = Forward(dummy); }
+        catch { /* best-effort; a real forward failure surfaces on the actual Train/Predict */ }
```
Comment (ooples) [inline src/Document/VisionLanguage/UDOP.cs]: _🩺 Stability & Availability_ | _🟠 Major_  | _⚡ Quick win_

**BLOCKING: Catch block discards the exception without logging it**

An empty catch makes the failure invisible to developers. Log it via Logger.LogError/LogWarning with the exception object, and surface only a sanitized message to users.

Flagged by `AIDN076` (Stability & Availability). This is a **regression**: the rule does not fire on this file on `master`, so this PR introduces it.

File: src/NER/Options/SpanBasedNEROptions.cs [hunk 2]
```diff
@@ -118,6 +118,32 @@ public int MaxSequenceLength
...
+    /// <summary>
+    /// Number of layers in each boundary FFNN that produces the span start/end representations.
+    /// </summary>
+    /// <remarks>
+    /// Yu et al.'s released configuration (juntaoy/biaffine-ner, experiments.conf) sets
+    /// <c>ffnn_depth = 2</c> alongside <c>ffnn_size = 150</c>. A single layer halves the depth of
+    /// the transformation that separates a token's start role from its end role, which is the
+    /// distinction the two separate FFNNs exist to draw.
+    /// </remarks>
+    public int FfnnDepth { get; set; } = 2;
```
Comment (ooples) [inline src/NER/Options/SpanBasedNEROptions.cs:130]: _🐛 Correctness_ | _🔴 Critical_  | _⚡ Quick win_

**BLOCKING: Copy constructor of 'SpanBasedNEROptions' does not copy property 'FfnnDepth'**

A copy constructor that misses a property is silent data loss: the clone keeps the default while the original keeps the configured value. This is the bug class behind the Tacotron2 and TimeBridge clone failures, where a clone loss stayed pinned at 306.7585 while the original fell to 0.2921.

Flagged by `AIDN070` (Correctness). This is a **regression**: the rule does not fire on this file on `master`, so this PR introduces it.

File: src/NER/Options/SpanBasedNEROptions.cs [hunk 3]
```diff
@@ -118,6 +118,32 @@ public int MaxSequenceLength
...
+
+    /// <summary>
+    /// Multiplicative learning-rate decay applied every <see cref="DecayFrequency"/> steps.
+    /// </summary>
+    /// <remarks>
+    /// Yu et al.'s released configuration (juntaoy/biaffine-ner, experiments.conf) sets
+    /// <c>decay_rate = 0.999</c>. The paper itself does not mention a schedule.
+    /// </remarks>
+    public double DecayRate { get; set; } = 0.999;
```
Comment (ooples) [inline src/NER/Options/SpanBasedNEROptions.cs:139]: _🐛 Correctness_ | _🔴 Critical_  | _⚡ Quick win_

**BLOCKING: Copy constructor of 'SpanBasedNEROptions' does not copy property 'DecayRate'**

A copy constructor that misses a property is silent data loss: the clone keeps the default while the original keeps the configured value. This is the bug class behind the Tacotron2 and TimeBridge clone failures, where a clone loss stayed pinned at 306.7585 while the original fell to 0.2921.

Flagged by `AIDN070` (Correctness). This is a **regression**: the rule does not fire on this file on `master`, so this PR introduces it.

File: src/NER/Options/SpanBasedNEROptions.cs [hunk 4]
```diff
@@ -118,6 +118,32 @@ public int MaxSequenceLength
...
+
+    /// <summary>
+    /// How many optimizer steps between applications of <see cref="DecayRate"/>.
+    /// </summary>
+    /// <remarks>The reference configuration uses <c>decay_frequency = 100</c>.</remarks>
+    public int DecayFrequency { get; set; } = 100;
```
Comment (ooples) [inline src/NER/Options/SpanBasedNEROptions.cs:145]: _🐛 Correctness_ | _🔴 Critical_  | _⚡ Quick win_

**BLOCKING: Copy constructor of 'SpanBasedNEROptions' does not copy property 'DecayFrequency'**

A copy constructor that misses a property is silent data loss: the clone keeps the default while the original keeps the configured value. This is the bug class behind the Tacotron2 and TimeBridge clone failures, where a clone loss stayed pinned at 306.7585 while the original fell to 0.2921.

Flagged by `AIDN070` (Correctness). This is a **regression**: the rule does not fire on this file on `master`, so this PR introduces it.

**Review by ooples — COMMENTED** [review]
**Actionable comments posted: 4**

File: src/ComputerVision/OCR/EndToEnd/ABCNet.cs [hunk 5]
```diff
@@ -0,0 +1,844 @@
...
+    public ABCNet(
+        ABCNetOptions<T> options,
+        NeuralNetworkArchitecture<T>? architecture = null,
+        IGradientBasedOptimizer<T, Tensor<T>, Tensor<T>>? optimizer = null,
+        ILossFunction<T>? lossFunction = null)
+        : base(ResolveArchitecture(options, architecture), lossFunction ?? new MeanSquaredErrorLoss<T>())
+    {
+        _options = options;
+        _lossFunction = lossFunction ?? new MeanSquaredErrorLoss<T>();
+        _optimizer = optimizer ?? new AdamOptimizer<T, Tensor<T>, Tensor<T>>(this);
```
Comment (ooples) [inline src/ComputerVision/OCR/EndToEnd/ABCNet.cs]: _🐛 Correctness_ | _🟠 Major_ | _⚡ Quick win_

**BLOCKING: the training rate is neither the paper's nor user-overridable.**

The optimizer is constructed bare, so it runs on AdamW's own generic default rather than a rate chosen for this model. I checked the options class and there is no `LearningRate` property at all, so a caller has no way to correct it either — the only escape is injecting a fully-built optimizer.

Both halves of the project's rule are broken here: defaults should be the **paper's**, and every default should stay **user-overridable**.

This is a regression — `master` does not construct this optimizer bare.

Fix: add `LearningRate` to the options class with the paper's value as its default, then read it here. I have not asserted a specific number because I could not verify this model's paper in this pass; take it from the paper's training section.

<details>
<summary>🛡️ Proposed fix: read the rate from options</summary>

```diff
-        _optimizer = optimizer ?? new AdamWOptimizer<T, Tensor<T>, Tensor<T>>(this);
+        _optimizer = optimizer ?? new AdamWOptimizer<T, Tensor<T>, Tensor<T>>(
+            this,
+            new AdamWOptimizerOptions<T, Tensor<T>, Tensor<T>>
+            {
+                InitialLearningRate = _options.LearningRate,
+            });
```

</details>

**Review by ooples — COMMENTED** [review]
**Actionable comments posted: 1**

File: src/NER/Options/SpanBasedNEROptions.cs (hunk 2 above)
Comment (ooples) [reply src/NER/Options/SpanBasedNEROptions.cs:130]: Already addressed on the current head. The copy constructor at `SpanBasedNEROptions.cs:267-288` copies this property (lines 283-285 cover `FfnnDepth`, `DecayRate` and `DecayFrequency`). This comment was raised against an earlier commit; verified against the branch tip before resolving.

File: src/NER/Options/SpanBasedNEROptions.cs (hunk 3 above)
Comment (ooples) [reply src/NER/Options/SpanBasedNEROptions.cs:139]: Already addressed on the current head. The copy constructor at `SpanBasedNEROptions.cs:267-288` copies this property (lines 283-285 cover `FfnnDepth`, `DecayRate` and `DecayFrequency`). This comment was raised against an earlier commit; verified against the branch tip before resolving.

File: src/NER/Options/SpanBasedNEROptions.cs (hunk 4 above)
Comment (ooples) [reply src/NER/Options/SpanBasedNEROptions.cs:145]: Already addressed on the current head. The copy constructor at `SpanBasedNEROptions.cs:267-288` copies this property (lines 283-285 cover `FfnnDepth`, `DecayRate` and `DecayFrequency`). This comment was raised against an earlier commit; verified against the branch tip before resolving.

File: src/Document/VisionLanguage/UDOP.cs (hunk 1 above)
Comment (ooples) [reply src/Document/VisionLanguage/UDOP.cs]: Fixed in d8096e38. The catch now retains the exception on the instance rather than discarding it.

Not logged, because there is nowhere to log to: AiDotNet is a library and carries no logger — nothing under `src/Document` references `ILogger`, `Debug.WriteLine` or any equivalent. The warm-up is best-effort by design and its failure must not propagate, so keeping the exception where a debugger, a test or a later diagnosis can reach it is the available way to stop the first and cleanest failure being invisible.

File: src/ComputerVision/OCR/EndToEnd/ABCNet.cs (hunk 5 above)
Comment (ooples) [reply src/ComputerVision/OCR/EndToEnd/ABCNet.cs]: Confirmed as a real defect, and **left open** because it cannot be fixed on this branch.

`ABCNetOptions<T>` is referenced throughout `ABCNet.cs` but its declaration exists nowhere in this slice — `git grep 'class ABCNetOptions'` returns nothing here, and it is absent from `split/1789-01-core-harness` and `split/1789-10-computervision-video` as well. So there is no options class on this branch to add `LearningRate` to.

That is the same cross-slice dependency documented in #1990: slices 01 and 03 define types each other consumes, and no merge order of the split produces a compiling tree. This fix needs whichever slice owns `ABCNetOptions`.

I have also deliberately not guessed the paper's learning rate. ABCNet (Liu et al., CVPR 2020) trains under Detectron2 with SGD, while this constructor builds an Adam optimizer, so the paper's SGD rate is not transferable to it. Setting a number I cannot justify would satisfy the analyzer while breaking the rule it enforces.

File: src/NER/Options/SpanBasedNEROptions.cs (hunk 2 above)
Comment (ooples) [reply src/NER/Options/SpanBasedNEROptions.cs:130]: Already fixed on this branch, in `828a9355a` ("fix(options): copy every property in options copy constructors") — `SpanBasedNEROptions.cs:283-285` carries `FfnnDepth`, `DecayRate` and `DecayFrequency`.

I re-audited the whole copy constructor rather than only the flagged line: all 7 auto-properties and all 11 backing fields are copied, `OnnxOptions` is deep-copied via its own copy constructor, and `LabelNames` is cloned rather than shared — so no property is carried by reference either.

File: src/NER/Options/SpanBasedNEROptions.cs (hunk 3 above)
Comment (ooples) [reply src/NER/Options/SpanBasedNEROptions.cs:139]: Already fixed on this branch, in `828a9355a` ("fix(options): copy every property in options copy constructors") — `SpanBasedNEROptions.cs:283-285` carries `FfnnDepth`, `DecayRate` and `DecayFrequency`.

I re-audited the whole copy constructor rather than only the flagged line: all 7 auto-properties and all 11 backing fields are copied, `OnnxOptions` is deep-copied via its own copy constructor, and `LabelNames` is cloned rather than shared — so no property is carried by reference either.

File: src/NER/Options/SpanBasedNEROptions.cs (hunk 4 above)
Comment (ooples) [reply src/NER/Options/SpanBasedNEROptions.cs:145]: Already fixed on this branch, in `828a9355a` ("fix(options): copy every property in options copy constructors") — `SpanBasedNEROptions.cs:283-285` carries `FfnnDepth`, `DecayRate` and `DecayFrequency`.

I re-audited the whole copy constructor rather than only the flagged line: all 7 auto-properties and all 11 backing fields are copied, `OnnxOptions` is deep-copied via its own copy constructor, and `LabelNames` is cloned rather than shared — so no property is carried by reference either.

File: src/Document/VisionLanguage/UDOP.cs (hunk 1 above)
Comment (ooples) [reply src/Document/VisionLanguage/UDOP.cs]: Fixed in `499ab7200`.

The catch is best-effort by design — a genuine forward failure surfaces again on the real `Train`/`Predict` — so it is still not rethrown. What was wrong is that `catch { }` discarded the diagnosis too, which meant the later failure carried no hint that the warm-up had already hit the same problem:

```csharp
try { _ = Forward(dummy); }
catch (Exception ex)
{
    System.Diagnostics.Debug.WriteLine(
        $"{nameof(UDOP<T>)}: lazy shape resolution failed on a {ImageSize}x{ImageSize} warm-up "
        + $"pass; shapes stay unresolved until the first real Train/Predict. {ex}");
}
```

File: src/ComputerVision/OCR/EndToEnd/ABCNet.cs (hunk 5 above)
Comment (ooples) [reply src/ComputerVision/OCR/EndToEnd/ABCNet.cs]: Fixed in `499ab7200`.

Both halves were real. `new AdamOptimizer<T, Tensor<T>, Tensor<T>>(this)` passed no options, so training fell through to `AdamOptimizerOptions.InitialLearningRate = 0.001`, and `ABCNetOptions` had no rate at all — so injecting a fully-built optimizer really was the only escape.

The property could not go in this PR: `ABCNetOptions<T>` lives at `src/Models/Options/ABCNetOptions.cs`, which this slice does not contain. It was added on **#1967 in `822f3db0c`**, and this commit consumes it:

```csharp
_optimizer = optimizer ?? new AdamOptimizer<T, Tensor<T>, Tensor<T>>(
    this,
    new AdamOptimizerOptions<T, Tensor<T>, Tensor<T>>
    {
        InitialLearningRate = _options.LearningRate,
    });
```

**One caveat, stated rather than buried.** You asked for the paper's value and the default is now the paper's `0.01` — but ABCNet (arXiv:2002.10200) trains with **SGD + momentum**, and this model defaults to **Adam**, where 0.01 is roughly two orders of magnitude above the usual scale. The published number and the default optimizer are not matched to each other, and no single default can satisfy both. Rather than pick silently, the remark on the property and the comment here both say so outright: inject an SGD-with-momentum optimizer to reproduce the paper, or lower this to ~1e-4 to train with the default Adam. `Validate` also now rejects a non-positive or non-finite rate.

Build: 36 errors before, 36 after, identical — all pre-existing split artifacts.
